File: Packets/src/main/java/me/ImSpooks/core/packets/init/PacketRegister.java
package me.ImSpooks.core.packets.init;

import me.ImSpooks.core.packets.collection.database.*;
import me.ImSpooks.core.packets.collection.network.PacketConfirmConnection;
import me.ImSpooks.core.packets.collection.network.PacketRequestConnection;
import me.ImSpooks.core.packets.collection.other.PacketPing;
import me.ImSpooks.core.packets.collection.other.PacketPingResponse;
import me.ImSpooks.core.packets.collection.other.PacketResponseExpired;
import me.ImSpooks.core.packets.type.PacketType;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Nick on 01 okt. 2019.
 * Copyright © ImSpooks
 */
public class PacketRegister {
    private static final Map<Integer, Class<? extends Packet>> REGISTERED_PACKETS = new HashMap<>();
    private static final Map<Class<? extends Packet>, Integer> REGISTERED_IDS = new HashMap<>();
    private static final Map<Integer, PacketType> PACKET_TYPES = new HashMap<>();

    private final static int MAX_PACKETS = (Short.MAX_VALUE + 1) * 2;

    static {
        // networking
        register(1, PacketRequestConnection.class, PacketType.NETWORK);
        register(2, PacketConfirmConnection.class, PacketType.NETWORK);
        register(3, PacketResponseExpired.class, PacketType.NETWORK);

        // database=
        register(1, PacketRequestData.class, PacketType.DATABASE);
        register(2, PacketRequestDataResponse.class, PacketType.DATABASE);
        register(3, PacketRequestCollection.class, PacketType.DATABASE);
        register(4, PacketRequestCollectionResponse.class, PacketType.DATABASE);
        register(5, PacketUpdateData.class, PacketType.DATABASE);

        //other
        register(1, PacketPing.class, PacketType.OTHER);
        register(2, PacketPingResponse.class, PacketType.OTHER);
    }

    private static void register(int id, Class<? extends Packet> packet, PacketType packetType) {
        id = packetType.START_ID + id - 1;

        if (REGISTERED_PACKETS.containsKey(id)) {
            throw new IllegalArgumentException(String.format("Packet with ID %s already registered for type %s", id - packetType.START_ID + 1, packetType));
        }
        if (REGISTERED_IDS.containsKey(packet)) {
            throw new IllegalArgumentException("Packet " + packet + " already registered");
        }
        if (REGISTERED_PACKETS.size() > MAX_PACKETS) {
            throw new IllegalArgumentException(String.format("Cannot register packet \'%s\': The maximum packet limit (%s) has been reached", packet.getSimpleName(), MAX_PACKETS));
        }

        REGISTERED_PACKETS.put(id, packet);
        REGISTERED_IDS.put(packet, id);
        PACKET_TYPES.put(id, packetType);
    }

    public static Packet createInstance(int id) {
        if (id < 0 || id > MAX_PACKETS)
            throw new IllegalArgumentException("Illegal id range " + id);

        try {
            Class<? extends Packet> p = REGISTERED_PACKETS.get(id);
            if (p == null)
                throw new IllegalArgumentException("Unknown packet ID " + id);

            Constructor<? extends Packet> s = p.getDeclaredConstructor();
            s.setAccessible(true);
            return s.newInstance();
        } catch (NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
    }

    public static int getId(Packet packet) {
        if (packet == null)
            throw new IllegalArgumentException("Packet may not be null");

        Integer id = REGISTERED_IDS.get(packet.getClass());
        if (id == null)
            throw new IllegalArgumentException("Unknown packet ID " + packet.getClass().getName());
        if (id < 0 || id > MAX_PACKETS)
            throw new AssertionError("Packet id had impossible value " + id);

        return id;
    }

    public static PacketType getPacketType(int id) {
        if (id < 0 || id > MAX_PACKETS && !String.valueOf(id).startsWith("-"))
            throw new IllegalArgumentException("Illegal id range " + id);
        if (!PACKET_TYPES.containsKey(id))
            throw new IllegalArgumentException("Invalid packet id " + id);
        return PACKET_TYPES.get(id);
    }

    public static String getPacketName(int id) {
        if (id < 0 || id > MAX_PACKETS && !String.valueOf(id).startsWith("-"))
            throw new IllegalArgumentException("Illegal id range " + id);
        if (!REGISTERED_PACKETS.containsKey(id))
            throw new IllegalArgumentException("Invalid packet id " + id);

        return REGISTERED_PACKETS.get(id).getSimpleName();
    }

    public static Class<? extends Packet> getPacketFromClassName(String className) throws ClassNotFoundException {
        for (Class<? extends Packet> packet : getPackets()) {
            if (packet.getSimpleName().equalsIgnoreCase(className)) {
                return packet;
            }
        }
        throw new ClassNotFoundException("No registered packets with name \'" + className + "\' found");
    }

    public static List<Class<? extends Packet>> getPackets() {
        return new ArrayList<>(REGISTERED_IDS.keySet());
    }
}

File: ClientBridge/src/main/php/client/packets/init/PacketRegister.php
<?php
/**
 * Created by Nick on 15 okt. 2019.
 * Copyright © ImSpooks
 */

namespace client\packets\init;

use AssertionError;
use client\packets\collection\other\PacketPing;
use client\packets\collection\other\PacketPingResponse;
use client\packets\type\PacketType;

use client\packets\collection\database\PacketRequestCollection;
use client\packets\collection\database\PacketRequestCollectionResponse;
use client\packets\collection\database\PacketRequestData;
use client\packets\collection\database\PacketRequestDataResponse;
use client\packets\collection\database\PacketUpdateData;
use client\packets\collection\other\PacketResponseExpired;
use InvalidArgumentException;
use stdClass;

class PacketRegister {
    // TODO convert to 7.4 if released
    /** @var array */ private static $REGISTERED_PACKETS = [];
    /** @var array */ private static $REGISTERED_IDS = [];
    /** @var array */ private static $PACKET_TYPES = [];

    private const MAX_PACKETS = 65536;

    public static function init() {
        //network
        self::register(3, PacketResponseExpired::class, PacketType::NETWORK);

        // database
        self::register(1, PacketRequestData::class, PacketType::DATABASE);
        self::register(2, PacketRequestDataResponse::class, PacketType::DATABASE);
        self::register(3, PacketRequestCollection::class, PacketType::DATABASE);
        self::register(4, PacketRequestCollectionResponse::class, PacketType::DATABASE);
        self::register(5, PacketUpdateData::class, PacketType::DATABASE);

        //other
        self::register(1, PacketPing::class, PacketType::OTHER);
        self::register(2, PacketPingResponse::class, PacketType::OTHER);
    }

    private static function register(int $id, string $packet, array $packetType) {
        $id = $packetType["START_ID"] + $id - 1;

        if (array_key_exists($id, self::$REGISTERED_PACKETS)) {
            throw new InvalidArgumentException(sprintf("Packet with ID %s already registered for type %s", $id - $packetType["START_ID"] + 1, $packetType["NAME"]));
        }
        if (array_key_exists($packet, self::$REGISTERED_IDS)) {
            throw new InvalidArgumentException(sprintf("Packet %s already registered.", $packet));
        }
        if (count(self::$REGISTERED_PACKETS) > self::MAX_PACKETS) {
            throw new InvalidArgumentException(sprintf("Cannot register packet \'%s\': The maximum packet limit (%s) has been reached", $packet, self::MAX_PACKETS));
        }

        self::$REGISTERED_PACKETS[$id] = $packet;
        self::$REGISTERED_IDS[$packet] = $id;
        self::$PACKET_TYPES[$id] = $packetType;
    }

    public static function createInstance(int $id): Packet {
        if ($id < 0 || $id > self::MAX_PACKETS)
            throw new InvalidArgumentException("Illegal id range " . $id);

        if (!array_key_exists($id, self::$REGISTERED_PACKETS))
            throw new InvalidArgumentException("Unknowk packet ID " . $id);

        $classname = self::$REGISTERED_PACKETS[$id];

        return new $classname();
    }

    public static function getId(string $packet): int {
        if ($packet == null)
            throw new InvalidArgumentException("Packet may not be null");

        if (!array_key_exists($packet, self::$REGISTERED_IDS))
            throw new InvalidArgumentException("Unknown packet ID " . $packet);

        $id = self::$REGISTERED_IDS[$packet];

        if ($id < 0 || $id > self::MAX_PACKETS)
            throw new AssertionError("Packet id had impossible value " . $id);

        return $id;
    }

    public static function getPacketType(int $id): array {
        if ($id < 0 || $id > self::MAX_PACKETS)
            throw new InvalidArgumentException("Illegal id range " . $id);
        if (!array_key_exists($id, self::$PACKET_TYPES))
            throw new InvalidArgumentException("Invalid packet ID " . $id);
        return self::$PACKET_TYPES[$id];
    }

    public static function getPacketName(int $id): string {
        if ($id < 0 || $id > self::MAX_PACKETS)
            throw new InvalidArgumentException("Illegal id range " . $id);
        if (!array_key_exists($id, self::$REGISTERED_PACKETS))
            throw new InvalidArgumentException("Invalid packet ID " . $id);
        return self::$REGISTERED_PACKETS[$id];
    }

    public static function getPackets(): array {
        return array_values(self::$REGISTERED_IDS);
    }
}